torch/csrc/api/src/optim/adagrad.cpp

```diff
@@ -11,42 +12,142 @@
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(int64_t, step);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(Tensor, sum);
+}
 
 /// Adapted from
 /// https://github.com/pytorch/pytorch/blob/master/torch/optim/adagrad.py
 void Adagrad::step() {
-  for (size_t i = 0; i < parameters_.size(); ++i) {
-    Tensor p = parameters_.at(i);
-    if (!p.grad().defined()) {
-      continue;
-    }
+  for (auto& group : param_groups_) {
+    for (auto& p : group.params()) {
+      if (!p.grad().defined()) {
+        continue;
+      }
+      auto grad = p.grad().data();
+      TORCH_INTERNAL_ASSERT(state_[c10::guts::to_string(p.unsafeGetTensorImpl())] != nullptr, "state found NULL for the Tensor ", p);
+      auto& state = static_cast<AdagradParamState&>(*state_[c10::guts::to_string(p.unsafeGetTensorImpl())]);
+      auto& options = static_cast<AdagradOptions&>(group.options());
+
+      state.step(state.step() + 1);
 
-    if (options.weight_decay() > 0) {
-      NoGradGuard guard;
-      p.grad() = p.grad() + options.weight_decay() * p;
+      if (options.weight_decay() != 0) {
+        TORCH_CHECK(!p.grad().data().is_sparse(), "weight_decay option is not compatible with sparse gradients");
+        grad = grad.add(p.data(), options.weight_decay());
+      }
+      const auto clr = options.lr() /
+          (1 + (state.step() - 1) * options.lr_decay());
+
+      if (grad.is_sparse()) {
+        grad = grad.coalesce();
+        auto grad_indices = grad._indices();
+        auto grad_values = grad._values();
+        auto size = grad.sizes();
+
+        auto make_sparse = [&] (const Tensor& values) -> Tensor {
+          if (grad_indices.dim() == 0 || values.dim() == 0) {
+            return torch::empty({0}, grad.options()).resize_as_(grad);
+          }
+          return torch::sparse_coo_tensor(grad_indices, values, size, grad.options());
+        };
+        state.sum(state.sum().add_(make_sparse(grad_values.pow(2))));
+        auto std = state.sum().sparse_mask(grad);
+        const auto std_values = std._values().sqrt_().add_(options.eps());
+
+        p.data().add_(make_sparse(grad_values / std_values), -clr);
+      }
+      else {
+        state.sum(state.sum().addcmul_(grad, grad, 1.0));
+        const auto std = state.sum().sqrt().add_(options.eps());
+        p.data().addcdiv_(grad, std, -clr);
+      }
     }
+  }
+}
 
-    buffer_at(step_buffers, i) += 1.0;
-    const auto clr = options.learning_rate() /
-        (1.0 + (buffer_at(step_buffers, i) - 1.0) * options.lr_decay());
+void Adagrad::add_parameters(const std::vector<Tensor>& parameters) {
+  param_groups_.emplace_back(OptimizerParamGroup(parameters, defaults_->clone()));
+}
 
-    auto& sum = buffer_at(sum_buffers, i);
-    sum.addcmul_(p.grad(), p.grad(), 1.0);
-    const auto std = buffer_at(sum_buffers, i).sqrt().add_(1e-10);
+const std::vector<Tensor>& Adagrad::parameters() const noexcept {
+  return param_groups_[0].params();
```

Comment:
It would be great to use `param_groups_.at(0)` instead of `param_groups_[0]`, because `.at` also does boundary checks.